test/T04Market.t.sol

```diff
@@ -260,21 +304,195 @@ contract T04MarketTest is D03ProtocolDefaults, MockAccounts {
+    }
+
+    function testOfferValidation() public {
+        testStartTokenSale();
+
+        vm.startPrank(account9);
+        vm.expectRevert("must belong to entity to make an offer");
+        nayms.executeLimitOffer(nWETH, dt.entity1MintAndSaleAmt, entity1, dt.entity1MintAndSaleAmt);
+        vm.stopPrank();
+
+        // init taker entity
+        nayms.createEntity(entity2, signer2Id, initEntity(weth, collateralRatio_500, maxCapital_2000eth, totalLimit_2000eth, true), "entity test hash");
+        nayms.externalDepositToEntity(entity2, wethAddress, 1_000 ether);
+
+        vm.startPrank(signer2);
+
+        vm.expectRevert("sell amount exceeds uint128 limit");
+        nayms.executeLimitOffer(nWETH, 2**128 + 1000, entity1, dt.entity1MintAndSaleAmt);
+
+        vm.expectRevert("buy amount exceeds uint128 limit");
+        nayms.executeLimitOffer(nWETH, dt.entity1MintAndSaleAmt, entity1, 2**128 + 1000);
+
+        vm.expectRevert("sell amount must be >0");
+        nayms.executeLimitOffer(nWETH, 0, entity1, dt.entity1MintAndSaleAmt);
+
+        vm.expectRevert("buy amount must be >0");
+        nayms.executeLimitOffer(nWETH, dt.entity1MintAndSaleAmt, entity1, 0);
+
+        vm.expectRevert("sell token must be valid");
+        nayms.executeLimitOffer("", dt.entity1MintAndSaleAmt, entity1, dt.entity1MintAndSaleAmt);
+
+        vm.expectRevert("buy token must be valid");
+        nayms.executeLimitOffer(nWETH, dt.entity1MintAndSaleAmt, "", dt.entity1MintAndSaleAmt);
+
+        vm.expectRevert("must be one platform token"); // 2 non-platform tokens
+        nayms.executeLimitOffer(nWETH, dt.entity1MintAndSaleAmt, nWBTC, dt.entity1MintAndSaleAmt);
+
+        vm.expectRevert("cannot sell and buy same token");
+        nayms.executeLimitOffer(nWETH, dt.entity1MintAndSaleAmt, nWETH, dt.entity1MintAndSaleAmt);
+
+        nayms.executeLimitOffer(nWETH, dt.entity1MintAndSaleAmt - 10 ether, entity1, dt.entity1MintAndSaleAmt);
+
+        vm.expectRevert("tokens locked in market");
+        nayms.executeLimitOffer(nWETH, dt.entity1MintAndSaleAmt, entity1, dt.entity1MintAndSaleAmt);
+
+        uint256 lastOfferId = nayms.getLastOfferId();
+        nayms.cancelOffer(lastOfferId);
+
+        vm.stopPrank();
+
+        nayms.startTokenSale(entity2, dt.entity2MintAndSaleAmt, dt.entity2SalePrice);
+
+        vm.startPrank(signer3);
+        vm.expectRevert("must be one platform token"); // 2 platform tokens
+        nayms.executeLimitOffer(entity2, dt.entity1MintAndSaleAmt, entity1, dt.entity1MintAndSaleAmt);
+
+        vm.stopPrank();
+    }
+
+    function testMatchingExternalTokenOnSellSide() public {
+        nayms.addSupportedExternalToken(wethAddress);
+        writeTokenBalance(account0, naymsAddress, wethAddress, dt.entity1StartingBal);
+
+        nayms.createEntity(entity1, signer1Id, initEntity(weth, collateralRatio_500, maxCapital_2000eth, totalLimit_2000eth, true), "entity test hash");
+
+        // start nENTITY1 token sale
```

Comment:
Am thinking if we do this AFTER we've setup the counter-offer then we should be able to test this line: https://coveralls.io/builds/52959033/source?filename=src%2Fdiamonds%2Fnayms%2Flibs%2FLibMarket.sol#L269

Thoughts?

Reply:
That is exactly the reason why the other nEntity1 sale is created afterwards on line 449, we have a problem with the offer assertion which does not allow to make an offer for an entity which does not exist. See [here](https://github.com/nayms/contracts-v3/blob/market_tests/src/diamonds/nayms/libs/LibMarket.sol#L354-L357). This test **does** trigger the else statement in `_buy` method, you can confirm by adding a log statement to it.

Reply:
Ok, so in that case we can create the entity1 before but then only do its token sale after the entity2 limit order has been created?

Reply:
Test goes like this:
1. Entity 1 starts a sale of 1000 "Entity 1" tokens.
2. Counter offer is placed to buy 2000 "Entity 1" tokens, half of which is matched immediately. Commissions are paid from the buying (taker) side.
3. Entity 1 starts another token sale for another 1000 "Entity 1" tokens. Sale offer is immediately matched to the offer from step 2, which is at this time partially filled. This time commissions are charged from the sell (maker) side.

But you are right, we can just create the entity before making an offer to buy it's tokens, we don't need two sales, we can skip the first token sale.

Reply:
Yeah good point. I'm then still left wondering why that line isn't being covered.